src/iotivity/iotivity_client.cc

```diff
@@ -0,0 +1,463 @@
+** You may obtain a copy of the License at
+** http://www.apache.org/licenses/LICENSE-2.0
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+**
+** $CISCO_END_LICENSE$
+**
+****************************************************************************/
+#include <string>
+#include <map>
+
+#include "iotivity/iotivity_client.h"
+#include "iotivity/iotivity_device.h"
+#include "iotivity/iotivity_resource.h"
+
+
+IotivityClient::IotivityClient(IotivityDevice* device) : m_device(device) {
+    m_asyncCallId_findresources = 0;
+    m_asyncCallId_finddevices = 0;
+    m_discoveryOptionDeviceId = "";
+}
+
+IotivityClient::~IotivityClient() {
+    for (auto const &entity : m_resourcemap) {
+        IotivityResourceClient *resClient =
+            reinterpret_cast<IotivityResourceClient *>(entity.second);
+        delete resClient;
+    }
+    m_resourcemap.clear();
+    m_foundresourcemap.clear();
+}
+
+void *IotivityClient::getResourceById(std::string id) {
+    if (m_resourcemap.size()) {
+        std::map<std::string, void *>::const_iterator it;
+        if ((it = m_resourcemap.find(id)) != m_resourcemap.end())
+            return reinterpret_cast<void *>((*it).second);
+    }
+
+    return NULL;
+}
+
+void IotivityClient::foundResourceCallback(std::shared_ptr<OCResource> resource,
+        int waitsec) {
+    DEBUG_MSG("foundResourceCallback:\n");
+    bool mapFoundResource = true;
+
+    IotivityResourceClient *resClient = new IotivityResourceClient(m_device);
+    resClient->setSharedPtr(resource);
+    PrintfOcResource((const OCResource &)*resource);
+
+    if (m_discoveryOptionDeviceId != "") {
+        if (resource->sid() != m_discoveryOptionDeviceId) {
+            mapFoundResource = false;
+        }
+    }
+
+    if (mapFoundResource) {
+    m_foundresourcemap[resClient->getResourceId()] =
+        reinterpret_cast<void *>(resClient);
+    }
+
+    if (waitsec == -1) {
+        findPreparedRequest();
+    }
+}
+
+void IotivityClient::handleFindResources(const picojson::value& value) {
+    DEBUG_MSG("handleFindResources: v=%s\n", value.serialize().c_str());
+
+    double async_call_id = value.get("asyncCallId").get<double>();
+    picojson::value param = value.get("OicDiscoveryOptions");
+    m_asyncCallId_findresources = async_call_id;
+
+    // all properties are null by default, meaning “find all”
+    // if resourceId is specified in full form, a direct retrieve is made
+    // if resourceType is specified, a retrieve on /oic/res is made
+    // if resourceId is null, and deviceId not,
+    // then only resources from that device are returned
+    std::string deviceId = "";
+    std::string resourceId = "";
+    std::string resourceType = "";
+    int waitsec = 5;
```

Comment:
This workaround with timeout is needed since it seems that the C++ resource discovery API is broken (see https://jira.iotivity.org/browse/IOT-678). I would use the C API for discovery (both resource and device), or wait until the C++ API is fixed? It should not be blocker for this PR, but this solution should really be temporary...